client-core/src/service/wallet_service.rs

```diff
@@ -774,9 +781,36 @@ where
         let info_keyspace = get_info_keyspace(name);
         let staking_count: u64 =
             read_number(&self.storage, &info_keyspace, "stakingkeyindex", None)?;
-        for i in 0..staking_count {
-            let pubkey = read_pubkey(&self.storage, &stakingkey_keyspace, &format!("{}", i))?;
-            ret.insert(pubkey);
+
+        if reversed {
+            let mut start = 0;
+            if staking_count > offset {
+                start = staking_count - 1 - offset;
+            }
+            let mut end = 0;
+            if limit > 0 && start >= (limit - 1) {
+                end = start - (limit - 1);
+            }
+            let mut i = start;
+            while i >= end {
+                let pubkey = read_pubkey(&self.storage, &stakingkey_keyspace, &format!("{}", i))?;
+                ret.insert(pubkey);
+                if 0 == i {
+                    break;
+                }
+                // iterate
+                i -= 1;
+            }
```

Comment:
I'm thinking if this could be simplified with something like:
```
let start = std::cmp::max(0, staking_count.saturating_sub(1).saturating_sub(offset));
let end = std::cmp::max(0, start.saturating_sub(limit.saturating_sub(1));
(start..end).map(|i| read_pubkey(&self.storage, &stakingkey_keyspace, &format!("{}", i))?).collect()
```

Reply:
looking good to me, i'll replace